vllm/sequence.py

```diff
@@ -119,6 +119,7 @@ def __init__(
             output_token_ids = []
 
         self.prompt_token_ids = prompt_token_ids
```

Comment:
do we need to store it as both list and a tuple?

Reply:
Technically we do not have to, as long as there is no code that directly change the value of `prompt_token_ids` attribute inside the class `SequenceGroup`.
I did a code search that confirms that there is no such code inside vLLM repo now. The search keyword is ".prompt_token_ids"
![image](https://github.com/vllm-project/vllm/assets/16879550/24d1e570-1d94-412e-842a-c9121bb47e9e)
And there is no such code that changes `prompt_token_ids` besides the initialization function.

Reply:
We should make it a tuple then, IMO. This also signifies prompt tokens are immutable.

Reply:
However, most of the existing codes are assuming the attribute `prompt_token_ids` to be of type `List[int]`, and only the hashing function requires it to be `Tuple[int]` so that it is immutable and thus hashable. From this perspective, I guess it is worthwhile to have an extra copy of `prompt_token_ids` so that both hashing is fast, and developers can still treat `prompt_token_ids` as a list.

Reply:
How about we at least make the attribute with list private and only allow access (and not setting) through a property? Not perfect since it can still be modified in place but better than nothing.

Another idea is to subclass list to create `FrozenList` with mutable methods set to raise an exception but that's a lot more complex.

Reply:
Since immutable prompt token IDs is the most important assumption, I'd also suggest to change it to tuple directly. It seems not an issue to me to change all List[int] to Tuple[int, ...]. At least type annotation should not be a blocker.

Reply:
I have changed `prompt_token_ids` to `Tuple[int]`, and fixed typing conflicts caused by this change.